text/0089-tenant-tokens.md

```diff
@@ -0,0 +1,298 @@
+- Title: Tenant Tokens
+- Start Date: 2021-10-15
+- Specification PR: [#89](https://github.com/meilisearch/specifications/pull/89)
+- Discovery Issue: [#51](https://github.com/meilisearch/product/issues/51)
+
+# Tenant Token
+
+## 1. Functional Specification
+
+### 1.1 Summary
+
+A `Tenant token` is generated on the user code to be used by an end-user. It allows users to have multi-tenant indexes and thus restricts access to documents depending on the end-user making the search request given enforced rules specified from the user business logic.
+
+A Tenant Token is a JWT containing the information necessary for MeiliSearch to verify it and extract permission/rules to apply it to the end user's search.
+
+#### 1.1.1 Summary Key Points
+
+- `Tenant tokens` are JWTs generated on the user side. Thus not stored or retrievable on MeiliSearch side.
+- `Tenant tokens` contain rules that ensure that a `Tenant token` holder (e.g an end-user) only has access documents matching rules chosen at the `tenant token` creation.
+- `Tenant tokens` are signed from a MeiliSearch `API key` on the user's code.
+- `Tenant tokens` cannot be less restrictive than the signing `API key` and can only be used for searching. A Tenant Token cannot search within more indexes than the API Key that signed that Tenant Token.
+- `Tenant tokens` can have different rules for each index accessible by the signing API key. These filters rule per index are described in an `searchRules` json object.
+- The only rule at the moment is the search parameter `filter`. Other rules may be added in the future.
+- When a request is made with the Tenant Token, MeiliSearch checks if this Tenant Token is authorized to make the request and in this case injects the rules at search time.
```

Comment:
Can we make this more explicit @gmourier? Because reading that now, makes me think in two possibilities:

- the SDK will decode the JWT and then get the `searchRules` to send them into the search request
- the Meilisearch server will receive the tenant token and it should be responsible to decode + apply these search rules on the fly.

Reply:
I've added some precisions. Let me know if it's enough. 😇